metabase/metabase
# Soft-delete defective duplicate fields

Pursuant towards https://github.com/metabase/metabase/issues/44659
Follows https://github.com/metabase/metabase/pull/44866 which populates `metabase_field`.`is_defective_duplicate`

Implements step 3 in [this comment](https://github.com/metabase/metabase/issues/44659#issuecomment-2192932023):

> 3. Create a migration to deactivate [defective duplicate] fields. After this point, users can not use defective fields in new questions.

This PR soft-deletes (by setting `active=false`) the subset of defective duplicate fields whose usage can be fully replaced by other fields without affecting the result of any query. It does not attempt to automatically remove these fields.

**To verify:**

1. Start from [[sync] parse naked json values #43812](https://github.com/metabase/metabase/pull/43812) with `git checkout 33df115`
2. Create a database with a few tables, JSON column and at least one other column. (The reason why I created multiple tables is because there is undeterministic behaviour in which the defective field can be synced either before or after the actual field. Depending on the order, the base type of the fields will be either `:type/JSON` or `:type/Array`.)
```sql
CREATE TABLE table_1 (id serial PRIMARY KEY, data jsonb);
INSERT INTO table_1 (data) VALUES ('[1, 2, 3]'::jsonb);
```
3. Create a question that selects only the Data column, and not ID column.
4. Switch to this branch `git switch soft-delete-is-defective-duplicate-fields` and restart Metabase.
5. Visit the question in step 3 and observe the question works as before and can be repaired manually.
